medrxivr blogpost (#57)
ropensci/roweb3: content/blog/2020-10-20-searching-medrxivr-and-biorxiv-preprint-data/index.Rmd

The primary aim of medrxivr is to make it as easy as possible to get cleaned preprint metadata into R. This enables a range of exploratory analysis to be readily performed. For example, to explore the distribution of topic categories for those preprints posted to medRxiv in the first week of 2020, the following code can be used:

``` {r categories, eval = FALSE}
# Load the packages
library(medrxivr)
library(dplyr)
library(ggplot2)

# Import the data from medRxiv API endpoint
mx_data <- mx_api_content(from_date = "2020-01-01", to_date = "2020-01-07")

# Create the graph 
mx_data %>%
  group_by(category) %>%
  summarise(N = n(), .groups = "keep") %>%
  arrange(desc(N)) %>%
  ggplot(aes(y = reorder(category,N), x = N)) +
  geom_col(colour = "black") + 
  ylab("Category") +
  xlab("Number of preprints, 1st week of January 2020") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,10)) +
  theme_classic() 
```

<!--html_preserve-->
{{< figure src = "categories-1.png" width = "600" alt = "Distribution of preprints posted to medRxiv in the first week of 2020 across topic categories." >}}
<!--/html_preserve-->

### Use case #2 – Reproducible searching and exporting

> While being able to easily import medRxiv data is useful, as a systematic reviewer, my main interest is in what medrxivr allows you to do once you have imported a local copy of the database. As mentioned in the section above, the native medRxiv website search interface is neither reproducible nor transparent, and while the API (accessed via `mx_api_content()`) is great for accessing medRxiv metadata, it does not have offer any search functionality. The second core aspect of medrxivr's functionality is designed to address these limitations, allowing information specialists to build complex searches and apply them to a local copy of the medRxiv database, all while documenting their search strategy in a transparent manner. An example is given below, where a researcher is looking for records that contain both dementia-related and lipid-related terms:

Reviewer: ```suggestion
While being able to easily import medRxiv data is useful, as a systematic reviewer, my main interest is in what medrxivr allows you to do once you have imported a local copy of the database. As mentioned in the section above, the native medRxiv website search interface is neither reproducible nor transparent, and while the API (accessed via `mx_api_content()`) is great for accessing medRxiv metadata, it does not offer any search functionality. The second core aspect of medrxivr's functionality is designed to address these limitations, allowing information specialists to build complex searches and apply them to a local copy of the medRxiv database, all while documenting their search strategy in a transparent manner. An example is given below, where a researcher is looking for records that contain both dementia-related and lipid-related terms:
```

> ``` {r}

Reviewer: Below I suggest printing the results, if you like that idea, then I also suggest increasing the width, just a bit to show more of the table
```suggestion
``` {r, R.options=list(width=100)}
```

# Use the maintained snapshot to quickly load today's copy of the medRxiv database

> # Note - this `preprint_data` object could be saved as a CSV in your project repository to aid reproducibility

Reviewer: Makes so much more sense!
Small edit to make it fit :smile:

```suggestion
# Note - this `preprint_data` object could be saved as a CSV in your project 
#        repository to aid reproducibility
```

preprint_data <- mx_snapshot()

# Build your search
topic1  <- c("dementia","vascular","alzheimer's")  # Combined with Boolean OR
topic2  <- c("lipids","statins","cholesterol")     # Combined with Boolean OR
myquery <- list(topic1, topic2)                    # Combined with Boolean AND

# Run your search
results <- mx_search(data = preprint_data,
                     query = myquery)

> 

Reviewer: Perhaps print the results to show the awesome tibble that is returned
```suggestion
results
```

```

Once you have run your search, exporting the results to a .BIB file for import into a reference manager, such as Zotero or Mendeley, is as simple as passing the results object to the `mx_export()` function.

Similarly, as screening the full text PDFs of records returned by a search against eligibility criteria forms a key part of a systematic review, a second helper function downloads the PDF for each record returned by `mx_search()`. Again, it is simply a case of passing the `results` object to the `mx_download()` function.

## Conclusion

> medRxiv is a fantastic resource and has been a key source of information related to the COVID-19 pandemic. However, some key issues with the website mean that its native search functionality precludes its use in systematic reviews. The medrxivr R package seeks to address these limitations by providing a user-friendly way to import and systematically search medRxiv and bioRxiv records in R.  Full documentation of the package functionality, particularly around the implementation of complex search strategies using syntax such as wildcards and the NEAR operator, which is used to find co-located terms (e.g. "cholesterol NEAR2 test" find records where "cholesterol" and "test" are separated by two or less words), is available from the [medrxivr website](https://docs.ropensci.org/medrxivr/). Finally, I wanted to note that during development, medrxivr benefited immensely from an [rOpenSci peer review](https://github.com/ropensci/software-review/issues/380) by [Tuija Sonkkila](https://github.com/tts) and [Najko Jahn](https://github.com/njahn82).

Reviewer: I had no idea there was such thing as a NEAR operator. That is SO cool

Author: Yeah, you can currently use it in `medrxivr` via some pretty ugly regex syntax, but implementing it in a nicer fashion is definitely on the horizon!

### Confusing information about Boolean operators

A key limitation of the native search on the medRxiv website comes from the use of Boolean operators in search strings. Boolean operators (*AND*, *OR*, *NOT*) are used by librarians to combine individual search terms into specific complex search strings. For example, "(dementia OR alzheimer)" would be expected to return any record containing either "dementia" or "alzheimer". Combining this with a second topic via *AND*, e.g. "(dementia OR alzheimer) AND (lipid)", should return any record containing either "dementia" or "alzheimer" once it also contains the term "lipid".

One of the key assumptions of Boolean operators is that the order of the terms does not matter, i.e. that "(dementia OR alzheimer) AND lipids" is equivalent to "(alzheimer OR dementia) AND lipids". However, running these searches on the 13th Oct 2020, the first gives 133 results and the second gives 126. It is not clear what is happening behind the scenes, but there is potentially some weighting of terms based on proximity. This means that key records may be missed based on the arbitrary ordering of the search terms.

<br>

### Lack of reproducbility

> One of the key tenets of systematic reviews is that literature searches are transparent and reproducible - that is, that a certain search strategy (e.g. "(dementia OR alzheimer) AND lipids") will provide the same number of results each time it is performed. However, anecdotal evidence from information specialists and medical librarians suggests this is not the case for the medRxiv native search:

Reviewer: ```suggestion
One of the key tenets of systematic reviews is that literature searches are transparent and reproducible - that is, that a certain search strategy (e.g. "(dementia OR alzheimer) AND lipids") will provide the same number of results each time it is performed. However, anecdotal evidence from information specialists and medical librarians suggests that medRxiv's websearch returns inconsistent citations over time.
```

<!--html_preserve-->
{{< tweet 1293973175244656641 >}}
<!--/html_preserve-->

<br>

### No batch export

Any systematic reviewer will tell you that for a database to be useful as an information source, the one thing it **needs** to have is an option to export the results of a given search _en masse_. At present, this is not something the medRxiv website supports:

<!--html_preserve-->
{{< tweet 1244905414019428353 >}}
<!--/html_preserve-->

<!--html_preserve-->
{{< tweet 1255485844783325185 >}}
<!--/html_preserve-->

> This means that records returned by a search must be exported by hand in small batches, an error-prone and time-consuming (not to mention anger-inducing!) method of extracting citations from a bibliographic database.

Reviewer: Absolutely fine to say "anger-inducing" here :grin:, I'm sure it is!

### Use Case #1 – Get preprint metadata in R

The primary aim of `medrxivr` is to make it as easy as possible to get cleaned preprint metadata into R. This enables a range of exploratory analysis to be readily performed. For example, to explore the distribution of topic categories for those preprints posted to medRxiv in the first week of 2020, the following code can be used:

``` {r categories, eval = FALSE}
# Import the data from the 
mx_data <- mx_api_content(from_date = "2020-01-01", to_date = "2020-01-07")

# Create the graph 
mx_data %>%
  group_by(category) %>%
  summarise(N = n(), .groups = "keep") %>%
  arrange(desc(N)) %>%
  ggplot(aes(y = reorder(category,N), x = N)) +
  geom_col(colour = "black") + 
  ylab("Category") +
  xlab("Number of preprints, 1st week of January 2020") +
  scale_x_continuous(expand = c(0, 0), limits = c(0,10)) +
  theme_classic() 
```

<!--html_preserve-->
{{< figure src = "categories-1.png" width = "600" alt = "Distribution of preprints posted to medRxiv in the first week of 2020 across topic categories." >}}
<!--/html_preserve-->

### Use Case #2 – Perform transparent and reproducible searches as part of a literature review, and export results

> While being able to easily import medRxiv data is useful, as a systematic reviewer, my main interest is in what `medrxivr` allows you to do once you have imported it. As mentioned in the section above, the native medRxiv website search interface is neither reproducible nor transparent. The second core aspect of `medrxivr`'s functionality is designed to address this, allowing information specialists to build complex searches and find records contained in the medRxiv databases that match the criteria, all while documenting their search strategy. An example is given below, where a researcher is looking for records that contain both dementia-related and lipid-related terms:

Author: So your second interpretation is correct - to search for relevant articles, you first have to import a local copy of the _whole_ database, which is then passed as the `data` argument to `mx_search()`.

I've added some text to this section to try and make this point clearer!

``` {r}

# Use the maintained snapshot to quickly load today's copy of the medRxiv database
# Note - this could be saved in your project repo to ensure reproducibility
preprint_data <- mx_snapshot()

# Build your search
topic1  <- c("dementia","vascular","alzheimer's")  # Combined with Boolean OR
topic2  <- c("lipids","statins","cholesterol")     # Combined with Boolean OR
myquery <- list(topic1, topic2)                    # Combined with Boolean AND

# Run your search
results <- mx_search(data = preprint_data,
                     query = myquery)

```

Once you have run your search, exporting the results to a .BIB file for import into a reference manager, such as Zotero or Mendeley, is a simply as passing the results object to the `mx_export()` function.

Similarly, as screening through the full text PDFs of records returned by a search against eligibility criteria forms a key part of a systematic review, a second helper function downloads the PDF for each record returned by `mx_search()`. Again, it is simply a case of passing the `results` object to the `mx_download()` function.

## Conclusion

> medRxiv is a fantastic resource and has been a key source of information related to the COVID-19 pandemic. However, some key issues with the website mean that its native search functionality precludes its use in systematic reviews. The `medrxivr` R package seeks to address these limitations by providing a user-friendly way to import and systematically search medRxiv and bioRxiv records in R. Full documentation on the functionality, particularly around the implementation of complex search strategies using syntax such as wildcards and NEAR, is available from the [package website](https://docs.ropensci.org/medrxivr/).

Author: I've added some text to explain the NEAR operator, but please let me know if it is still unclear!

> medRxiv is a fantastic resource and has been a key source of information related to the COVID-19 pandemic. However, some key issues with the website mean that its native search functionality precludes its use in systematic reviews. The `medrxivr` R package seeks to address these limitations by providing a user-friendly way to import and systematically search medRxiv and bioRxiv records in R. Full documentation on the functionality, particularly around the implementation of complex search strategies using syntax such as wildcards and NEAR, is available from the [package website](https://docs.ropensci.org/medrxivr/).

Author: I've added a linkto the review thread, and added both reviewers names/links to GH profiles to the text.
